Add wrapped object factory test

brodybits opened the pull request:
Similar to the example at: https://nodejs.org/api/addons.html#addons_factory_of_wrapped_objects

kkoopa commented on test/cpp/wrappedobjectfactory.cpp:
```diff
@@ -0,0 +1,70 @@
+
+    SetPrototypeMethod(tpl, "getValue", GetValue);
+
+    constructor.Reset(tpl->GetFunction());
```
`GetFunction(tpl).ToLocalChecked()`

kkoopa commented on test/cpp/wrappedobjectfactory.cpp:
```diff
@@ -0,0 +1,70 @@
+
+  static NAN_METHOD(NewInstance) {
+    v8::Local<v8::Function> cons = Nan::New(constructor);
+    double value = info[0]->IsUndefined() ? 0 : info[0]->NumberValue();
```
`To<double>(info[0]).FromJust()`

kkoopa commented on test/cpp/wrappedobjectfactory.cpp:
```diff
@@ -0,0 +1,70 @@
+    v8::Local<v8::Function> cons = Nan::New(constructor);
+    double value = info[0]->IsUndefined() ? 0 : info[0]->NumberValue();
+    v8::Local<v8::Value> argv[1] = {Nan::New(value)};
+    info.GetReturnValue().Set(cons->NewInstance(1, argv));
```
`NewInstance(const, 1, argv).ToLocalChecked()`

kkoopa commented on test/cpp/wrappedobjectfactory.cpp:
```diff
@@ -0,0 +1,70 @@
+
+  static NAN_METHOD(New) {
+    if (info.IsConstructCall()) {
+      double value = info[0]->IsUndefined() ? 0 : info[0]->NumberValue();
```
`To<double>(info[0]).FromJust()`

kkoopa commented on test/cpp/wrappedobjectfactory.cpp:
```diff
@@ -0,0 +1,70 @@
+      const int argc = 1;
+      v8::Local<v8::Value> argv[argc] = {info[0]};
+      v8::Local<v8::Function> cons = Nan::New(constructor);
+      info.GetReturnValue().Set(cons->NewInstance(argc, argv));
```
`NewInstance(const, 1, argv).ToLocalChecked()`

kkoopa commented on test/cpp/wrappedobjectfactory.cpp:
```diff
@@ -0,0 +1,70 @@
+  MyObject::Init(target);
+  Set(target
+    , New<v8::String>("newFactoryObjectInstance").ToLocalChecked()
+    , New<v8::FunctionTemplate>(MyObject::NewInstance)->GetFunction()
```
`GetFunction(New<v8::FunctionTemplate>(MyObject::NewInstance)).ToLocalChecked()`

kkoopa commented:
Thanks, looks good apart from future-proofing.

brodybits commented:
I pushed some changes but please do not merge it yet! It fails when I try to run it on node 0.10.40, will investigate. Closing for now, maybe will raise a new PR.

kkoopa commented on test/cpp/wrappedobjectfactory.cpp:
```diff
@@ -0,0 +1,71 @@
+  );
+}
+
+NODE_MODULE(objectwraphandle, Init)
```
They failed because this module is not called objectwraphandle.

brodybits commented:
The problems should now be fixed, I tried it on Node 0.10.40 and 4.1.1 (was OK on Node 0.12.7 and 4.0.0 in the past). I added another change where the wrapped object can return another wrapped object (I need this functionality for my own project), and rebased on the latest master changes.

kkoopa commented:
Looks good. Test failure on travis seems to be a network error downloading something.

brodybits commented:
Travis aargh @#$%

Thanks for looking at this.

brodybits commented:
I pushed a couple more changes:

1. Small fix to `test/js/wrappedobjectfactory-test.js`:
```diff
--- a/test/js/wrappedobjectfactory-test.js
+++ b/test/js/wrappedobjectfactory-test.js
@@ -10,7 +10,7 @@ const test     = require('tap').test
     , testRoot = require('path').resolve(__dirname, '..')
     , bindings = require('bindings')({ module_root: testRoot, bindings: 'wrappedobjectfactory' });
 
-test('objectwraphandle', function (t) {
+test('wrappedobjectfactory', function (t) {
   t.plan(4);
 
   var obj = bindings.newFactoryObjectInstance(123);
```

2. Also fix `test/cpp/objectwraphandle.cpp` to move constructor object into a private static inline function

kkoopa commented:
Merged as 4392aa9df745960d553a47e61ae58d8bb811753c bff133ab771eef56315bb2f498a0d95777246f1f 3a4a0ae15bcf35e177d35087063146151c56449f

brodybits commented:
Thanks for merging my changes!
